// grunt-comment-toggler/tasks/lib/comment_toggler.js
/*
 * grunt-comment-toggler
 *
 *
 * Copyright (c) 2014 Kimmo Salmela
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

    var exports = {};

    exports.options = {
        removeCommands: false,
        padding: 1
    };

    // Format: <!-- comments:(action) (comment delimiter) -->
    var startRegex = /<!--\s*comments:\s*(\w*)\s*([^\s]+)\s*-->/i;
    var endRegex = /<!--\s*endcomments\s*-->/i;

    // This regex captures full comment blocks that start with "[command]-block"
    // command.
    // Capture groups:
    // 1 - command
    // 2 - delimiter
    // 3 - content
    var blockRegex = /.*<!--\s*comments:\s*(.*?)-block\s*([^\s]+)\s*-->.*\n([\s\S]*?)\n.*<!--\s*endcomments\s*-->.*/gi;

    var commands = {
        search: "search",
        comment: "comment",
        uncomment: "uncomment",
        toggle: "toggle"
    };

    var operation = {
        command: commands.search,
        startChar: "",
        endChar: ""
    };

    /**
     * Looks for comment build blocks inside content, processes them and returns
     * the processed content.
     *
     * @param {String} content
     * @returns {String}
     */
    exports.processFile = function(content) {
        // Normalize newlines and store the original newline character that
        // should be restored.
        var newlineChar = (content.match(/\r\n/g)) ? "\r\n" : "\n";
        content = content.replace(/\r\n/g, "\n");

        // Process block comments
        content = processBlockComments(content);

        // Split content to an array of lines to process line comments
        var lines = content.split("\n");
        var padding = exports.options.padding;

        // Process line comments
        lines.forEach(function(line, i, lines) {

            var cmd = operation.command;

            if (cmd !== commands.search && line.match(endRegex)) {
                operation.command = commands.search;

                if (exports.options.removeCommands) {
                    lines[i] = "";
                }
                return;
            }

            if (cmd === commands.search) {
                var m = line.match(startRegex);

                if (m) {
                    // Extract command from match
                    operation.command = commands[m[1].toLowerCase()];

                    // Extract delimiter from match and handle possible special
                    // delimiters
                    var delims = getSpecialDelimiters(m[2]);
                    operation.startChar = delims[0] || m[2];
                    operation.endChar = delims[1] || "";

                    if (exports.options.removeCommands) {
                        lines[i] = "";
                    }
                }

                return;
            }

            var processedLine = line;

            if (cmd === commands.comment || cmd === commands.toggle) {
                processedLine = comment(line, operation.startChar, operation.endChar, padding);
            }

            // If toggled line didn't change after trying to comment it it means
            // that it should be uncommented instead.
            if (cmd === commands.uncomment || (cmd === commands.toggle && processedLine === line)) {
                processedLine = uncomment(line, operation.startChar, operation.endChar);
            }

            lines[i] = processedLine;
        });

        // If we're still running a comment operation after the last line it
        // means that there was a missing "endcomments" tag
        if (operation.command !== commands.search) {
            grunt.log.warn("Missing 'endcomments' tag.");
        }

        // Join split lines and restore line endings
        content = lines.join("\n");
        content = content.replace(/\n/g, newlineChar);

        return content;
    };

    /**
     * Returns an array of comment start and end delimiters based on the
     * supplied special delimiter string.
     *
     * The first element of the returned array is the start delimiter and the
     * second one is the end delimiter. Currently supported special delimiters
     * are "html" and "css". Returns an empty array if the passed special
     * delimiter is not supported.
     *
     * @param {String} delimiterStr Special delimiter string
     * @returns {Array}
     */
    var getSpecialDelimiters = function(delimiterStr) {
        var delims = [];

        switch (delimiterStr.toLowerCase()) {
            case "html":
                delims[0] = "<!--";
                delims[1] = "-->";
                break;

            case "css":
                delims[0] = "/*";
                delims[1] = "*/";
                break;
        }

        return delims;
    };

    /**
     * Processes block comments and returns processed content
     *
     * @param {String} content
     * @returns {String}
     */
    var processBlockComments = function(content) {
        var match;

        var newContent = content;

        // Process each block comment command
        while ((match = blockRegex.exec(content)) !== null) {
            var full = match[0];
            var command = match[1].toLowerCase();
            var delimiter = match[2];
            var blockContent = match[3];

            var padding = exports.options.padding;

            var delims = getSpecialDelimiters(delimiter);
            var startChar = delims[0] || "";
            var endChar = delims[1] || "";

            var replacement = blockContent;

            if (command === "comment" || command === "toggle") {
                replacement = comment(blockContent, startChar, endChar, padding);
            }

            if (command === "uncomment" || (command === "toggle" && replacement === blockContent)) {
                replacement = uncomment(blockContent, startChar, endChar);
            }

            if (exports.options.removeCommands) {
                // Replace command blocks with newlines
                replacement = "\n" + replacement + "\n";
                newContent = newContent.replace(full, replacement);
            } else {
                // Replace the blockContent in the full match first before
                // replcaing the whole block with commands included in the
                // original content
                var replacedFull = full.replace(blockContent, replacement);
                newContent = newContent.replace(full, replacedFull);
            }
        }

        return newContent;
    };

    /**
     * Returns commented content
     *
     * Adds comment delimiters to the content and returns it.
     *
     * @param {String} content
     * @param {String} [startDelim=""] Comment start delimiter
     * @param {String} [endDelim=""] Comment end delimiter
     * @param {Number} [padding=0] Padding between line content and delimiters
     * @returns {String}
     */
    var comment = function(content, startDelim, endDelim, padding) {

        padding = padding || 0;

        // Regex to check and return already commented line
        var r = generateCommentRegex(startDelim, endDelim);
        if (content.match(r)) {
            return content;
        }

        // Separate whitespace from content to retain indenting and newlines
        // (whitespace) (content) (endWhitespace)
        var r2 = /^(\s*)([\s\S]*?)(\s*)$/g;
        var m = r2.exec(content);
        var startWs = m[1];
        var innerContent = m[2];
        var endWs = m[3];

        // Create a padding string with length according to options
        // TODO - ES6 - http://goo.gl/WFr5qY
        var paddingStr = new Array(padding + 1).join(" ");

        // Setup delimiter strings
        startDelim = startDelim || "";
        // Add padding to end delimiter if it's used
        endDelim = (endDelim) ? paddingStr + endDelim : "";

        // Build the commented line
        content = startWs + startDelim + paddingStr + innerContent.trim() + endDelim + endWs;

        return content;
    };

    /**
     * Removes comment delimiters from the content and returns it.
     *
     * Returns content without comment delimiters and with whitespace trimmed
     * between the delimiters and the content. Newlines are preserved.
     *
     * @param {String} content
     * @param {String} [startDelim=""] Comment start delimiter
     * @param {String} [endDelim=""] Comment end delimiter
     * @returns {String}
     */
    var uncomment = function(content, startDelim, endDelim) {

        startDelim = startDelim || "";
        endDelim = endDelim || "";

        // Capture: (whitespace) (content)
        var r = generateCommentRegex(startDelim, endDelim);
        var m = r.exec(content);

        // Regex tries to match comment characters and capture the inner
        // content. This is used to tell what should be returned.
        if (m === null) {
            return content;
        } else {
            // (startWs) (content) (endWs)
            return m[1] + m[2] + m[3];
        }
    };

    /**
     * Returns a comment RegExp constructed from start and end delimiters.
     *
     * The returned RegExp matches commented content and captures following
     * groups:
     * 1. Whitespace before comment delimiter
     * 2. Content between comment delimiters
     * 3. Whitespace after comment delimiter
     *
     * @param {String} startDelim
     * @param {String} endDelim
     * @returns {RegExp}
     */
    var generateCommentRegex = function(startDelim, endDelim) {
        // [^\S\n]* is used after and before delimiters instead of \s* to
        // preserve newlines and only trim line whitespace between content and
        // comment delimiters.
        // Regex explanation:
        // (whitespace) [delimiter] [whitespace (no \n)] (content) [whitepsace (no \n)] [delimiter] (whitespace)
        // Capture groups:
        // (whitespace) (content) (endWhitespace)
        var r = new RegExp("^(\\s*)" + escapeRegExp(startDelim) + "[^\\S\\n]*([\\s\\S]*?)[^\\S\\n]*" + escapeRegExp(endDelim) + "(\\s*)$", "g");
        //grunt.log.writeln(r);
        return r;
    };

    // http://goo.gl/L7cZZN
    var escapeRegExp = function(string) {
        return string.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1");
    };

    return exports;

};
