// grunt-comment-toggler/Gruntfile.js
/*
 * grunt-comment-toggler
 * https://github.com/Cryszon/grunt-comment-toggler
 *
 * Copyright (c) 2014 Kimmo Salmela
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

    // Project configuration.
    grunt.initConfig({
        jshint: {
            all: [
                'Gruntfile.js',
                'tasks/*.js',
                '<%= nodeunit.tests %>'
            ],
            options: {
                jshintrc: '.jshintrc'
            }
        },
        // Before generating any new files, remove any previously-created files.
        clean: {
            tests: ['tmp']
        },
        // Configuration to be run (and then tested).
        toggleComments: {
            default_options: {
                files: {
                    "tmp/default_options/blockComments": "test/fixtures/blockComments.fix",
                    "tmp/default_options/lineComments": "test/fixtures/lineComments.fix",
                    "tmp/default_options/specialLineComments": "test/fixtures/specialLineComments.fix",
                }
            },
            custom_options: {
                options: {
                    padding: 4,
                    removeCommands: true
                },
                files: {
                    "tmp/custom_options/blockComments": "test/fixtures/blockComments.fix",
                    "tmp/custom_options/lineComments": "test/fixtures/lineComments.fix",
                    "tmp/custom_options/specialLineComments": "test/fixtures/specialLineComments.fix",
                }
            }
        },
        // Unit tests.
        nodeunit: {
            tests: ['test/*_test.js']
        }

    });

    // Actually load this plugin's task(s).
    grunt.loadTasks('tasks');

    // These plugins provide necessary tasks.
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-nodeunit');

    // Whenever the "test" task is run, first clean the "tmp" dir, then run this
    // plugin's task(s), then test the result.
    grunt.registerTask('test', ['clean', 'toggleComments', 'nodeunit']);

    // By default, lint and run all tests.
    grunt.registerTask('default', ['jshint', 'test']);

};
